Released version either as source or docker repo that actually works

Hi

I am trying to stitch together an Fabric8 platform primarily for Java based applications with OpenShift and the Fabric8MQ and ActiveMQ based templates as I understand comes from this repository.

Though none of the recent versions appears to actually work. The most recent one 2.2.90 and the snapshot version 2.2.91-SNAPSHOT appears to fail on the basic start of the ActiveMQ broker.

org.jboss.weld.exceptions.DeploymentException: WELD-000123: Error loading io.fabric8.cdi.weld.ClientProducer defined in <class>io.fabric8.cdi.weld.ClientProducer</class> in jar:file:/deployments/lib/fabric8-cdi-2.2.91-tests.jar!/META-INF/beans.xml@21
        at org.jboss.weld.bootstrap.enablement.GlobalEnablementBuilder$ClassLoader.apply(GlobalEnablementBuilder.java:269)
        at org.jboss.weld.bootstrap.enablement.GlobalEnablementBuilder$ClassLoader.apply(GlobalEnablementBuilder.java:256)
        at com.google.common.collect.Lists$TransformingRandomAccessList.get(Lists.java:572)
        at java.util.AbstractList$Itr.next(AbstractList.java:358)
        at java.util.AbstractCollection.toArray(AbstractCollection.java:141)
        at com.google.common.collect.ImmutableSet.copyOf(ImmutableSet.java:378)
        at org.jboss.weld.bootstrap.enablement.GlobalEnablementBuilder.createModuleEnablement(GlobalEnablementBuilder.java:233)
        at org.jboss.weld.bootstrap.BeanDeployment.createEnablement(BeanDeployment.java:219)
        at org.jboss.weld.bootstrap.WeldStartup.startInitialization(WeldStartup.java:377)
        at org.jboss.weld.bootstrap.WeldBootstrap.startInitialization(WeldBootstrap.java:76)
        at org.jboss.weld.environment.se.Weld.initialize(Weld.java:146)
        at io.fabric8.amq.Main.main(Main.java:31)
Caused by: java.lang.NoClassDefFoundError: org/easymock/IAnswer
        at java.lang.Class.getDeclaredConstructors0(Native Method)
        at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
        at java.lang.Class.getDeclaredConstructors(Class.java:2020)
        at org.jboss.weld.environment.deployment.WeldResourceLoader.classForName(WeldResourceLoader.java:52)
        at org.jboss.weld.bootstrap.enablement.GlobalEnablementBuilder$ClassLoader.apply(GlobalEnablementBuilder.java:267)


Can you point to a tag in either the master branch or another branch that have "release quality" which I can trust to start from?


Best regards
Lars Milland

Hi

Thanx for the quick response - are you, as in James Strachan maybe also able to investigate a maybe similar problem with the Fabric8MQ classpath/dependency setup where a launch of the docker image from the build replies:

[vagrant@vagrant ~]$ docker run -i -t fabric8/fabric8mq:2.2.91-SNAPSHOT
I> No access restrictor found, access to all MBean is allowed
I> No suitable address found for listening on multicast discovery requests
Jolokia: Agent started with URL http://172.17.0.2:8778/jolokia/
Exception in thread "main" java.lang.NoClassDefFoundError: org/slf4j/LoggerFactory
        at io.fabric8.mq.Main.<clinit>(Main.java:24)
Caused by: java.lang.ClassNotFoundException: org.slf4j.LoggerFactory
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 1 more

--- a/amqbroker/pom.xml
+++ b/amqbroker/pom.xml
@@ -51,7 +51,6 @@
     <dependency>
       <groupId>io.fabric8</groupId>
       <artifactId>fabric8-cdi</artifactId>
-
     </dependency>
     <dependency>
       <groupId>org.apache.activemq</groupId>
@@ -83,42 +82,39 @@
       <artifactId>arquillian-junit-container</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
+    </dependency>
+<!--
+    TODO not sure why this dependency was here but it breaks hawt-app as it gets included and then breaks Weld
+
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>fabric8-cdi</artifactId>
+      <scope>test</scope>
+      <type>test-jar</type>
+    </dependency>
+-->
+    <dependency>
+      <groupId>org.jboss.arquillian.container</groupId>
+      <artifactId>arquillian-weld-ee-embedded-1.1</artifactId>
+      <version>1.0.0.CR8</version>
+      <scope>test</scope>
+    </dependency>
 
     <dependency>
       <groupId>io.fabric8</groupId>
       <artifactId>fabric8-arquillian</artifactId>
       <scope>test</scope>
     </dependency>
-
     <dependency>
       <groupId>io.fabric8</groupId>
       <artifactId>kubernetes-assertions</artifactId>
       <scope>test</scope>
     </dependency>
 
-    <dependency>
-      <groupId>junit</groupId>
-      <artifactId>junit</artifactId>
-      <scope>test</scope>
-    </dependency>
-
-    <dependency>
-      <groupId>org.jboss.weld</groupId>
-      <artifactId>weld-core</artifactId>
-      <version>2.2.5.Final</version>
-    </dependency>
-
-    <dependency>
-      <groupId>org.jboss.weld.se</groupId>
-      <artifactId>weld-se</artifactId>
-      <version>2.2.5.Final</version>
-    </dependency>
-
-    <dependency>
-      <groupId>org.jboss.arquillian.container</groupId>
-      <artifactId>arquillian-weld-ee-embedded-1.1</artifactId>
-      <version>1.0.0.CR8</version>
-    </dependency>
   </dependencies>
 
   <build>
@@ -155,7 +151,7 @@
       </plugin>
 
       <plugin>
-        <groupId>io.fabric8</groupId>
+        <groupId>org.jolokia</groupId>
         <artifactId>docker-maven-plugin</artifactId>
         <version>${docker.maven.plugin.version}</version>
         <configuration>
@@ -167,7 +163,6 @@
                 <assembly>
                   <basedir>/deployments</basedir>
                   <descriptorRef>hawt-app</descriptorRef>
-
                 </assembly>
                 <env>
                   <JAVA_LIB_DIR>/deployments/lib</JAVA_LIB_DIR>

--- a/amqbroker/pom.xml
+++ b/amqbroker/pom.xml
@@ -76,6 +76,20 @@
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-log4j12</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.jboss.weld</groupId>
+      <artifactId>weld-core</artifactId>
+<!--
+      <version>2.2.5.Final</version>
+-->
+    </dependency>
+    <dependency>
+      <groupId>org.jboss.weld.se</groupId>
+      <artifactId>weld-se</artifactId>
+<!--
+      <version>2.2.5.Final</version>
+-->
+    </dependency>
 
     <dependency>
       <groupId>org.jboss.arquillian.junit</groupId>

--- a/msg-gateway/src/main/java/io/fabric8/msg/gateway/brokers/impl/EndpointWatcher.java
+++ b/msg-gateway/src/main/java/io/fabric8/msg/gateway/brokers/impl/EndpointWatcher.java
@@ -1,0 +1,123 @@
+/**
+ * Copyright (C) 2016 Red Hat, Inc.
+ * <p/>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.msg.gateway.brokers.impl;
+
+import io.fabric8.kubernetes.api.model.EndpointSubset;
+import io.fabric8.kubernetes.api.model.Endpoints;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+
+
+/**
+ * Watches {@link Endpoints} objects
+ */
+public class EndpointWatcher implements Watcher<Endpoints> {
+    private static final Logger LOG = LoggerFactory.getLogger(KubernetesBrokerControl.class);
+    private static final Map<NamespaceName, Long> endpointsVersions = new HashMap<>();
+
+    private final KubernetesBrokerControl kubernetesBrokerControl;
+    private final String defaultNamespace;
+
+    public EndpointWatcher(KubernetesBrokerControl kubernetesBrokerControl, String defaultNamespace) {
+        this.kubernetesBrokerControl = kubernetesBrokerControl;
+        this.defaultNamespace = defaultNamespace;
+    }
+
+    @Override
+    public void onClose(KubernetesClientException e) {
+        if (e != null) {
+            LOG.warn(e.getMessage(), e);
+        }
+    }
+
+
+    public void onInitialEndpoints(Endpoints endpoints) throws IOException {
+        upsertEndpoint(endpoints);
+    }
+
+    @Override
+    public void eventReceived(Watcher.Action action, Endpoints endpoint) {
+        try {
+            switch (action) {
+                case ADDED:
+                    upsertEndpoint(endpoint);
+                    break;
+                case DELETED:
+                    deleteEndpoint(endpoint);
+                    break;
+                case MODIFIED:
+                    modifyEndpoint(endpoint);
+                    break;
+            }
+        } catch (IOException | InterruptedException e) {
+            LOG.warn("Caught: " + e, e);
+        }
+    }
+
+    private void upsertEndpoint(Endpoints endpoints) throws IOException {
+        synchronized (endpointsVersions) {
+            NamespaceName namespacedName = NamespaceName.create(endpoints);
+            Long resourceVersion = getResourceVersion(endpoints);
+            Long previousResourceVersion = endpointsVersions.get(namespacedName);
+
+            // lets only process this EndpointSubset if the resourceVersion is newer than the last one we processed
+            if (previousResourceVersion == null || (resourceVersion != null && resourceVersion > previousResourceVersion)) {
+                endpointsVersions.put(namespacedName, resourceVersion);
+
+                try {
+                    kubernetesBrokerControl.createClientForEndpoints(endpoints);
+                } catch (Exception e) {
+                    LOG.warn("Failed to update Endpoints " + namespacedName + ": " + e, e);
+                }
+            } else {
+                LOG.info("Ignored out of order notification for Endpoints " + namespacedName
+                        + " with resourceVersion " + resourceVersion + " when we have already processed " + previousResourceVersion);
+            }
+        }
+    }
+
+
+    private void modifyEndpoint(Endpoints endpoint) throws IOException, InterruptedException {
+        upsertEndpoint(endpoint);
+    }
+
+    private void deleteEndpoint(Endpoints endpoint) throws IOException, InterruptedException {
+        NamespaceName namespaceName = NamespaceName.create(endpoint);
+        synchronized (endpointsVersions) {
+            endpointsVersions.remove(namespaceName);
+        }
+    }
+
+    public static Long getResourceVersion(HasMetadata hasMetadata) {
+        ObjectMeta metadata = hasMetadata.getMetadata();
+        String resourceVersionText = metadata.getResourceVersion();
+        Long resourceVersion = null;
+        if (resourceVersionText != null && resourceVersionText.length() > 0) {
+            resourceVersion = Long.parseLong(resourceVersionText);
+        }
+        return resourceVersion;
+    }
+}

--- a/msg-gateway/src/main/java/io/fabric8/msg/gateway/brokers/impl/KubernetesBrokerControl.java
+++ b/msg-gateway/src/main/java/io/fabric8/msg/gateway/brokers/impl/KubernetesBrokerControl.java
@@ -16,12 +16,15 @@
 package io.fabric8.msg.gateway.brokers.impl;
 
 import io.fabric8.kubernetes.api.KubernetesHelper;
+import io.fabric8.kubernetes.api.model.DoneableEndpoints;
 import io.fabric8.kubernetes.api.model.EndpointAddress;
 import io.fabric8.kubernetes.api.model.EndpointPort;
 import io.fabric8.kubernetes.api.model.EndpointSubset;
 import io.fabric8.kubernetes.api.model.Endpoints;
 import io.fabric8.kubernetes.client.DefaultKubernetesClient;
 import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.dsl.ClientResource;
 import io.fabric8.kubernetes.client.utils.Utils;
 import io.fabric8.msg.gateway.ArtemisClient;
 import io.fabric8.msg.gateway.brokers.BrokerControl;
@@ -50,6 +53,8 @@ public class KubernetesBrokerControl implements BrokerControl {
     private ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
     private ConcurrentLinkedDeque<ArtemisClient> artemisClients = new ConcurrentLinkedDeque<>();
     private Map<Destination, ArtemisClient> destinationArtemisClientMap = new ConcurrentHashMap<>();
+    private EndpointWatcher endpointWatcher;
+    private Watch endpointWatch;
 
     @Override
     public ArtemisClient get(Destination destination) {
@@ -69,14 +74,26 @@ public class KubernetesBrokerControl implements BrokerControl {
     public void start() throws Exception {
         if (started.compareAndSet(false, true)) {
             kubernetesClient = new DefaultKubernetesClient();
-            executor.scheduleAtFixedRate(() -> lookupBrokers(), 0, 5, TimeUnit.SECONDS);
+
+            endpointWatcher = new EndpointWatcher(this, kubernetesClient.getNamespace());
+
+            ClientResource<Endpoints, DoneableEndpoints> endpointsClient;
+            endpointsClient = kubernetesClient.endpoints().inNamespace(getNamespace()).withName(getArtemisName());
+            endpointWatch = endpointsClient.watch(endpointWatcher);
+            Endpoints endpoints = endpointsClient.get();
+            if (endpoints != null) {
+                endpointWatcher.onInitialEndpoints(endpoints);
+            }
+
+            // TODO watch for the map of Destination -> Broker
+
             LOG.info("KubernetesBrokerControl started");
         }
     }
 
     public void stop() throws Exception {
-        if (started.compareAndSet(true, false)) {
-            executor.shutdownNow();
+        if (endpointWatch != null) {
+            endpointWatch.close();
         }
     }
 
@@ -120,55 +137,45 @@ public class KubernetesBrokerControl implements BrokerControl {
         this.portName = portName;
     }
 
-    protected void lookupBrokers() {
-        System.err.println("LOOKUP BROKERS CALLED");
-        try {
-
-            Endpoints endpoints = kubernetesClient.endpoints().inNamespace(getNamespace()).withName(getArtemisName()).get();
-    System.err.println("ENDPOINTS = " + endpoints);
-
-            HashSet<ArtemisClient> set = new HashSet<>();
-            if (endpoints != null) {
-                for (EndpointSubset subset : endpoints.getSubsets()) {
-                    if (subset.getPorts().size() == 1) {
-                        EndpointPort port = subset.getPorts().get(0);
-                        for (EndpointAddress address : subset.getAddresses()) {
-                            ArtemisClient artemisClient = new ArtemisClient(address.getIp(), port.getPort());
-                            set.add(artemisClient);
-                        }
-                    } else {
-                        for (EndpointPort port : subset.getPorts()) {
-                            if (Utils.isNullOrEmpty(portName) || portName.endsWith(port.getName())) {
-                                for (EndpointAddress address : subset.getAddresses()) {
-                                    ArtemisClient artemisClient = new ArtemisClient(address.getIp(), port.getPort());
-                                    set.add(artemisClient);
-                                }
+    public void createClientForEndpoints(Endpoints endpoints) throws Exception {
+        HashSet<ArtemisClient> set = new HashSet<>();
+        if (endpoints != null) {
+            for (EndpointSubset subset : endpoints.getSubsets()) {
+                if (subset.getPorts().size() == 1) {
+                    EndpointPort port = subset.getPorts().get(0);
+                    for (EndpointAddress address : subset.getAddresses()) {
+                        ArtemisClient artemisClient = new ArtemisClient(address.getIp(), port.getPort());
+                        set.add(artemisClient);
+                    }
+                } else {
+                    for (EndpointPort port : subset.getPorts()) {
+                        if (Utils.isNullOrEmpty(portName) || portName.endsWith(port.getName())) {
+                            for (EndpointAddress address : subset.getAddresses()) {
+                                ArtemisClient artemisClient = new ArtemisClient(address.getIp(), port.getPort());
+                                set.add(artemisClient);
                             }
                         }
                     }
                 }
             }
-
-            System.err.println("LOOKUP SET SIZE = " + set.size());
-
-            for (ArtemisClient artemisClient : set) {
-                if (!artemisClients.contains(artemisClient)) {
-                    artemisClient.start();
-                    artemisClients.add(artemisClient);
-                }
-            }
-
-            for (ArtemisClient artemisClient : artemisClients) {
-                if (!set.contains(artemisClient)) {
-                    artemisClient.stop();
-                    artemisClients.remove(artemisClient);
-                    LOG.info("Removed stale " + artemisClient);
-                }
-            }
-        } catch (Throwable e) {
-            LOG.error("lookupBrokers", e);
         }
 
+        System.err.println("LOOKUP SET SIZE = " + set.size());
+
+        for (ArtemisClient artemisClient : set) {
+            if (!artemisClients.contains(artemisClient)) {
+                artemisClient.start();
+                artemisClients.add(artemisClient);
+            }
+        }
+
+        for (ArtemisClient artemisClient : artemisClients) {
+            if (!set.contains(artemisClient)) {
+                artemisClient.stop();
+                artemisClients.remove(artemisClient);
+                LOG.info("Removed stale " + artemisClient);
+            }
+        }
     }
 
     private ArtemisClient getNextClient() {

--- a/msg-gateway/src/main/java/io/fabric8/msg/gateway/brokers/impl/NamespaceName.java
+++ b/msg-gateway/src/main/java/io/fabric8/msg/gateway/brokers/impl/NamespaceName.java
@@ -1,0 +1,79 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.msg.gateway.brokers.impl;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.utils.Objects;
+
+/**
+ * Represents a name in a namespace we can use as a key in a map
+ */
+public class NamespaceName {
+    private final String namespace;
+    private final String name;
+
+    public NamespaceName(String namespace, String name) {
+        this.namespace = namespace;
+        this.name = name;
+    }
+
+    public static NamespaceName create(HasMetadata hasMetadata) {
+        Objects.notNull(hasMetadata, "resource");
+        ObjectMeta metadata = hasMetadata.getMetadata();
+        Objects.notNull(metadata, "metadata");
+        String name = metadata.getName();
+        String namespace = metadata.getNamespace();
+        Objects.notNull(name, "metadata.name");
+        Objects.notNull(namespace, "metadata.namespace");
+        return new NamespaceName(namespace, name);
+    }
+
+    @Override
+    public String toString() {
+        return "NamespaceName{" +
+                namespace + ":" + name +
+                '}';
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        NamespaceName that = (NamespaceName) o;
+
+        if (!namespace.equals(that.namespace)) return false;
+        return name.equals(that.name);
+
+    }
+
+    @Override
+    public int hashCode() {
+        int result = namespace.hashCode();
+        result = 31 * result + name.hashCode();
+        return result;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public String getNamespace() {
+        return namespace;
+    }
+}

--- a/pom.xml
+++ b/pom.xml
@@ -74,7 +74,7 @@
         <integrationTestPattern>**/*KubernetesTest*.java</integrationTestPattern>
 
         <!-- upstream vs redhat versions of dependencies goes here -->
-        <fabric8.version>2.2.97</fabric8.version>
+        <fabric8.version>2.2.99</fabric8.version>
         <fabric8.release.version>${fabric8.version}</fabric8.release.version>
         <fabric8.devops.version>2.2.54</fabric8.devops.version>
         <apiman.version>1.2.3.CR1</apiman.version>

